File: Code/LinkedList.cpp
#include <bits/stdc++.h>
using namespace std;

struct node
{
    int info;
    struct node *link;

    node() {}
    node(int x)
    {
        info = x;
        link = NULL;
    }
};
struct node *start = NULL;

void insertion()
{
    int item;
    cout << "\nEnter data\n";
    cin >> item;
    struct node *n = new node;
    n->info = item;
    n->link = start;
    start = n;
}
void insertionAtEnd()
{
    int item;
    cout << "\nEnter data\n";
    cin >> item;
    struct node *n = new node;
    n->info = item;
    node *tmp = start;

    if (tmp == NULL)
    {
        n->link = NULL;
        start = n;
        return;
    }

    while (tmp->link != NULL)
        tmp = tmp->link;

    tmp->link = n;
    n->link = NULL;
}

void insertionAfterNode()
{
    int item;
    cout << "\nEnter data: ";
    cin >> item;
    struct node *n = new node();
    n->info = item;
    n->link = NULL;

    node *curr = start;
    node *prev = NULL;

    while (curr != NULL && curr->info <= item)
    {
        prev = curr;
        curr = curr->link;
    }
    // beginning insertion
    if (prev == NULL)
    {
        n->link = start;
        start = n;
    }
    else
    {
        // making new_node->next = prev_node->link
        n->link = prev->link;
        prev->link = n;
    }
}
void deleteNode()
{
    int item;
    cout << "Enter item to delete: ";
    cin >> item;
    node *ptr = start;
    if (ptr == NULL)
    {
        cout << "Underflow Error!\nNo Element to delete";
    }
    node *prev = NULL;
    node *curr = start;

    while (ptr != NULL && ptr->info != item)
    {
        prev = ptr;
        ptr = ptr->link;
    }

    if (ptr == NULL)
    {
        cout << "Element does not exist in the list\n";
    }
    else
    {
        if (prev == NULL)
        {
            start = ptr->link;
        }

        else
        {
            prev->link = ptr->link;
            free(ptr);
        }
    }
}
void display()
{
    struct node *ptr = start;

    if (ptr == NULL)
    {
        cout << "Linked List is Empty!";
        return;
    }
    cout << "\nElements are : ";
    while (ptr != NULL)
    {
        cout << ptr->info << "\t";
        ptr = ptr->link;
    }
}
int main()
{
    int c = 1;
    while ((c > 0) && (c < 6))
    {
        cout << "\n1. Insertion at the beg. \n2. Insertion at End \n3. Insertion after Node\n4. Delete node\n5. Display \n";
        cout << ">";
        cin >> c;
        switch (c)
        {
        case 1:
            insertion();
            break;
        case 2:
            insertionAtEnd();
            break;
        case 3:
            insertionAfterNode();
            break;
        case 4:
            deleteNode();
            break;
        case 5:
            display();
            break;
        default:
            cout << "\nWrong input..exiting..";
        }
    }
    return 0;
}

File: LinkedList.java
import java.util.Scanner;

class Node {

  int data;
  Node ptr;

  Node() {}

  Node(int x) {
    data = x;
    ptr = null;
  }
}

public class LinkedList {

  static Scanner sc = new Scanner(System.in);
  static Node start = null;

  public static void main(String[] args) {
    int c = 1;

    while (c > 0 && c < 3) {
      System.out.println("1. Insertion after Node\n2. Display \n");
      System.out.print(">");
      c = sc.nextInt();

      switch (c) {
        case 1:
          insertion();
          break;
        case 2:
          display();
          break;
      }
    }
  }

  static void insertion() {
    System.out.println("Enter data: ");
    int item = sc.nextInt();
    Node new_node = new Node(item);

    Node prev = null;
    Node curr = start;

    while (curr != null && curr.data <= item) {
      prev = curr;
      curr = curr.ptr;
    }

    if (prev == null) {
      new_node.ptr = start;
      start = new_node;
    } else {
      new_node.ptr = prev.ptr;
      prev.ptr = new_node;
    }
  }

  static void display() {
    Node temp = start;
    System.out.println("Elements are: ");
    while (temp != null) {
      System.out.print(temp.data);
      if (temp.ptr != null) {
        System.out.print(" -> ");
      }
      temp = temp.ptr;
    }
    System.out.println();
  }
}
